pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,353 @@
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+
+	tmplProcessor := template.NewProcessor(r.client, r.scheme)
+	objs, err := tmplProcessor.Process(tmplContent, params)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+	objs = template.Filter(objs, template.RetainAllButNamespaces)
+	err = tmplProcessor.Apply(objs)
+	if err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err,
+			"failed to provision namespace '%s'", nsName)
+	}
+
+	namespace := &corev1.Namespace{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: nsName}, namespace); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to get namespace '%s'", nsName)
+	}
+	if namespace.Labels == nil {
+		namespace.Labels = make(map[string]string)
+	}
+	namespace.Labels["revision"] = tcNamespace.Revision
+	if err := r.client.Update(context.TODO(), namespace); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to update namespace '%s'", nsName)
+	}
+
+	log.Info("namespace provisioned", "namespace", tcNamespace)
+	if err := r.setStatusProvisioning(nsTmplSet); err != nil {
+		return err
+	}
+
+	// TODO add validation for other objects
+	return nil
+}
+
+func nextMissingNamespace(tcNamespaces []toolchainv1alpha1.Namespace, namespaces []corev1.Namespace, username string) (*toolchainv1alpha1.Namespace, *corev1.Namespace, bool) {
+	for _, tcNamespace := range tcNamespaces {
+		nsName := ToNamespaceName(username, tcNamespace.Type)
+		namespace, found := findNamespace(namespaces, nsName)
+		if found {
+			if namespace.Status.Phase == corev1.NamespaceActive && namespace.GetLabels()["revision"] != tcNamespace.Revision {
+				return &tcNamespace, &namespace, true
+			}
+		} else {
+			return &tcNamespace, nil, true
+		}
+	}
+	return nil, nil, false
+}
+
+func findNamespace(namespaces []corev1.Namespace, namespaceName string) (corev1.Namespace, bool) {
+	for _, ns := range namespaces {
+		if ns.GetName() == namespaceName {
+			return ns, true
+		}
+	}
+	return corev1.Namespace{}, false
+}
+
+// ToNamespaceName returns NamespaceName formed using given userName and nsType
+func ToNamespaceName(userName, nsType string) string {
+	return fmt.Sprintf("%s-%s", userName, nsType)
+}
+
+// error handling methods
+
+func (r *ReconcileNSTemplateSet) wrapErrorWithStatusUpdate(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, statusUpdater func(nsTmplSet *toolchainv1alpha1.NSTemplateSet, message string) error, err error, format string, args ...interface{}) error {
```

Comment:
this func does 2 things which are not related, and it's confusing. It would be simpler to split the status update in the NSTemplateSet vs the error wrapping, IMO

Reply:
This is copied from `UserAccountCtrl` for consistency [link](https://github.com/codeready-toolchain/member-operator/blob/master/pkg/controller/useraccount/useraccount_controller.go#L318).  I guess status handling was well discussed at the time of UserAccountCtrl, we can revisit though.

Reply:
TL;DR `wrapErrorWithStatusUpdate()` is a de facto pattern we use in our controllers for error handling.

Long story. @xcoulon there are two kind of status updates.
One is a regular status update when we need to report the status like "Provisioning" or "Ready".
The other one is when we get an error. Among other things during the error handling we need to update the status (Ready=false, reason={reason}, message={error_message}). But the status update can fail too. And we don't want to hide the original error and return the status update error in the controller in this case.
We tried a few options but it turned out that a generic `wrapErrorWithStatusUpdate` function is pretty helpful for these kind of updates.
Splinting it up would make the code more messy.

But if you have ideas how we could improve it then I would encourage you to create a separate PR (for UserAccount ctrl for example). And if it's better than we have now we can adopt it everywhere.